src/main/java/org/cactoos/iterator/Matched.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017-2020 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.iterator;
+
+import java.util.Iterator;
+import org.cactoos.BiFunc;
+import org.cactoos.func.UncheckedBiFunc;
+import org.cactoos.text.FormattedText;
+
+/**
+ * Matched iterator.
+ *
+ * Iterates over 2 iterators at the same time and verify that elements with the
+ * same position has correlation by the function. The function might be equals,
+ * endsWith, greaterThen, nonNull, empty, negative, positive, etc.
+ *
+ * @param <X> Type of item in first iterator.
+ * @param <Y> Type of item in second iterator.
+ * @since 0.47
+ */
+public final class Matched<X, Y> implements Iterator<X> {
```

Comment:
@andreoss from previous discussions, I thought the Y parameter could be moved at the constructor level

Reply:
@victornoel It seems that only way to do so, is to create an anonymous `Iterator` within ctor and use `IteratorEnvelope`. Otherwise
since there is a field for `Func<X,Y>`, type parameter `Y` should go on class level

Reply:
@andreoss ok, let's forget about it then. I'm not 100% happy with this but I don't think introducing this complexity with the envelope is worth the effort :)